include/picolibrary/vector.h

```diff
@@ -0,0 +1,1326 @@
+    {
+        auto const n = end - begin;
+
+        auto element = this->begin() + ( begin - this->begin() );
+
+        for ( ; element != end; ++element ) { element->~Value(); } // for
+
+        for ( ; element != this->end(); ++element ) {
+            new ( element - n ) Value{ std::move( *element ) };
+
+            element->~Value();
+        } // for
+
+        m_size -= n;
+
+        return this->begin() + ( begin - this->begin() );
+    }
+
+    /**
+     * \brief Append an element to the end of the vector.
+     *
+     * \param[in] value The value to append to the vector.
+     *
+     * \return Nothing if the vector has sufficient capacity to store the new element.
+     * \return picolibrary::Generic_Error::INSUFFICIENT_CAPACITY if the vector does not
+     *         have sufficient capacity to store the new element.
+     */
+    constexpr auto push_back( Value const & value ) noexcept -> Result<Void, Error_Code>
+    {
+        if ( m_size + 1 > N ) {
+            return Generic_Error::INSUFFICIENT_CAPACITY;
+        } // if
+
+        new ( &m_storage[ m_size ] ) Value{ value };
+
+        ++m_size;
+
+        return {};
+    }
+
+    /**
+     * \brief Append an element to the end of the vector.
+     *
+     * \param[in] value The value to append to the vector.
+     *
+     * \return Nothing if the vector has sufficient capacity to store the new element.
+     * \return picolibrary::Generic_Error::INSUFFICIENT_CAPACITY if the vector does not
+     *         have sufficient capacity to store the new element.
+     */
+    constexpr auto push_back( Value && value ) noexcept -> Result<Void, Error_Code>
+    {
+        if ( m_size + 1 > N ) {
+            return Generic_Error::INSUFFICIENT_CAPACITY;
+        } // if
+
+        new ( &m_storage[ m_size ] ) Value{ std::move( value ) };
+
+        ++m_size;
+
+        return {};
+    }
+
+    /**
+     * \brief Append an element to the end of the vector.
+     *
+     * \tparam Arguments Element construction argument types.
+     *
+     * \param[in] arguments Element construction arguments.
+     *
+     * \return A reference to the appended element if the vector has sufficient capacity
+     *         to store the new element.
+     * \return picolibrary::Generic_Error::INSUFFICIENT_CAPACITY if the vector does not
+     *         have sufficient capacity to store the new element.
+     */
+    template<typename... Arguments>
+    constexpr auto emplace_back( Arguments &&... arguments ) noexcept
+        -> Result<std::reference_wrapper<Value>, Error_Code>
+    {
+        if ( m_size + 1 > N ) {
+            return Generic_Error::INSUFFICIENT_CAPACITY;
+        } // if
+
+        new ( &m_storage[ m_size ] ) Value{ std::forward<Arguments>( arguments )... };
+
+        ++m_size;
+
+        return back();
+    }
+
+    /**
+     * \brief Remove the last element of the vector.
+     */
+    constexpr void pop_back() noexcept
```

Comment:
This function can't be `constexpr` since it uses `reinterpret_cast`.
```suggestion
    void pop_back() noexcept
```